{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# seqFormatteR V2\n",
    "### Contact Information\n",
    "\n",
    "Anthony S. Castanza, PhD - acastanza@ucsd.edu\n",
    "### Summary\n",
    "https://github.com/ACastanza/seqformat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installation \n",
    "Install required R libraries for the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "install.packages(\"BiocManager\", repos=\"http://cran.r-project.org\")\n",
    "library(\"BiocManager\")\n",
    "BiocManager::install(c(\"GEOquery\",\"tximport\", \"GenomicFeatures\", \"biomaRt\", \"rhdf5\", \"DESeq2\"))\n",
    "install.packages(\"tidyverse\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Configure Enviornment\n",
    "Load required packages and configure global variables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "library(\"tools\")\n",
    "library(\"dplyr\")\n",
    "library(\"tidyr\")\n",
    "library(\"GEOquery\")\n",
    "library(\"tximport\")\n",
    "library(\"GenomicFeatures\")\n",
    "library(\"biomaRt\")\n",
    "library(\"DESeq2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (exists(\"altnorm\") == FALSE) {\n",
    "  altnorm <- FALSE\n",
    "}\n",
    "## before running script set altnorm <- 'userlog' or 'usevst' or 'useall' to\n",
    "## output additional normalize using deseq2's rlog or vst functions\n",
    "\n",
    "iscounts <- FALSE\n",
    "txlevel <- FALSE\n",
    "FAIL <- FALSE\n",
    "isnormalized <- FALSE\n",
    "TYPE <- NULL\n",
    "txtype <- 0\n",
    "is_directory <- FALSE\n",
    "getgeofiles <- FALSE\n",
    "seondaryfactor <- FALSE\n",
    "NORM <- FALSE\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import data files directly from GEO (optional)\n",
    "Directly import deposited data from a published experiment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "getgeofiles <- askYesNo(\"Attempt to get data directly from GEO \\\"supplementary files\\\"? \")\n",
    "if (getgeofiles == TRUE) {\n",
    "  geoid <- readline(prompt = (\"Enter the GEOID for the datafiles (eg: GSE38786): \"))\n",
    "  geooutfiles <- getGEOSuppFiles(geoid, makeDirectory = TRUE, baseDir = getwd(),\n",
    "    fetch_files = TRUE, filter_regex = NULL)\n",
    "  cat(\"Getting experiment information from Series Matrix...\\n\")\n",
    "  gsefile <- getGEO(geoid)\n",
    "  phenotypedata <- as.data.frame(pData(phenoData(gsefile[[1]]))[, c(\"geo_accession\",\n",
    "    \"title\")], stringsAsFactors = FALSE)\n",
    "  rownames(phenotypedata) <- 1:nrow(phenotypedata)\n",
    "  geostructure <- pData(phenoData(gsefile[[1]]))\n",
    "  cat(\"\\n\")\n",
    "  print(phenotypedata)\n",
    "  cat(\"\\n\")\n",
    "  cat(\"Files acquired from GEO:\\n\")\n",
    "  geoimporttable <- as.data.frame(rownames(geooutfiles), stringsAsFactors = FALSE)\n",
    "  print(rownames(geooutfiles))\n",
    "  # print('All')\n",
    "  cat(\"\\n\")\n",
    "  if (nrow(geoimporttable) > 1) {\n",
    "    useall <- askYesNo(\"Use all downloaded files? (\\\"No\\\" allows you to select a specific file)\")\n",
    "    if (useall == TRUE) {\n",
    "      genematrix <- paste0(getwd(), \"/\", geoid)\n",
    "    }\n",
    "    if (useall == FALSE) {\n",
    "      geoselected <- readline(prompt = (\"Select which GEO file to use for downstream processing: \"))\n",
    "      geoselectednumber <- match(geoselected, cbind(rownames(geoimporttable),\n",
    "        geoimporttable)[, 1])\n",
    "      if (is.na(geoselectednumber) == TRUE) {\n",
    "        geoselectednumber <- match(geoselected, cbind(rownames(geoimporttable),\n",
    "          geoimporttable)[, 2])\n",
    "      }\n",
    "      outfile <- rownames(geooutfiles)[geoselectednumber]\n",
    "      if (file_ext(rownames(geooutfiles)[geoselectednumber]) == \"tar\") {\n",
    "        untar(rownames(geooutfiles)[geoselectednumber], exdir = paste0(dirname(rownames(geooutfiles)[geoselectednumber]),\n",
    "          \"/\", basename(tools::file_path_sans_ext(rownames(geooutfiles)[geoselectednumber]))))\n",
    "        outfile <- paste0(dirname(rownames(geooutfiles)[geoselectednumber]),\n",
    "          \"/\", basename(tools::file_path_sans_ext(rownames(geooutfiles)[geoselectednumber])))\n",
    "      }\n",
    "      genematrix <- outfile\n",
    "    }\n",
    "  } else if (nrow(geoimporttable) == 1) {\n",
    "    if (file_ext(rownames(geooutfiles)[1]) == \"tar\") {\n",
    "      untar(rownames(geooutfiles)[1], exdir = paste0(dirname(rownames(geooutfiles)[1]),\n",
    "        \"/\", basename(tools::file_path_sans_ext(rownames(geooutfiles)[1]))))\n",
    "      outfile <- paste0(dirname(rownames(geooutfiles)[1]), \"/\", basename(tools::file_path_sans_ext(rownames(geooutfiles)[1])))\n",
    "    } else {\n",
    "      outfile <- paste0(geoimporttable[1])\n",
    "    }\n",
    "    genematrix <- outfile\n",
    "  }\n",
    "\n",
    "  cat(\"\\n\")\n",
    "  cat(\"Experiment Imported.\\n\")\n",
    "  cat(\"\\n\")\n",
    "\n",
    "  findcounts <- apply(geostructure, 2, function(x) {\n",
    "    grepl(\"counts|Counts\", x)\n",
    "  })\n",
    "  if (any(findcounts) == TRUE) {\n",
    "    message(\"Series Matrix implies that this data consists of COUNTS:\\n\")\n",
    "    print(unique(geostructure[findcounts]))\n",
    "    iscounts <- askYesNo(\"Do you agree that this is gene COUNTS data? \")\n",
    "    cat(\"\\n\")\n",
    "    if (iscounts == TRUE) {\n",
    "      countsdetected <- TRUE\n",
    "      istx <- FALSE\n",
    "      TYPE <- \"COUNTS\"\n",
    "    }\n",
    "  } else if (any(findcounts) == FALSE) {\n",
    "    cat(\"We couldn't automatically set the datatype\\n\")\n",
    "    cat(\"We'll prompt you to manually select datatype next.\\n\")\n",
    "  }\n",
    "  findtxquant <- apply(geostructure, 2, function(x) {\n",
    "    grepl(\"almon|ailfish|allisto\", x)\n",
    "  })\n",
    "  if (any(findtxquant) == TRUE) {\n",
    "    cat(\"\\n\")\n",
    "    message(\"Series Matrix implies that this data might be transcript level quantifications:\\n\")\n",
    "    print(unique(geostructure[findtxquant]))\n",
    "    cat(\"\\n\")\n",
    "    message(\"Validate the presence of transcript level quantifications in data files, then continue.\\n\")\n",
    "    cat(\"\\n\")\n",
    "  }\n",
    "\n",
    "  findnormal <- apply(geostructure, 2, function(x) {\n",
    "    grepl(\"normalized|Normalized|NORMALIZED|normalised|Normalised|NORMALISED\",\n",
    "      x)\n",
    "  })\n",
    "  if (any(findnormal) == TRUE) {\n",
    "    message(\"Series Matrix implies that this data might be ALREADY NORMALIZED:\\n\")\n",
    "    print(unique(geostructure[findnormal]))\n",
    "    isnormalized <- askYesNo(\"Is this data already normalized? \")\n",
    "    if (isnormalized == TRUE) {\n",
    "      NORM <- TRUE\n",
    "      DESEQ2DONE <- FALSE\n",
    "    }\n",
    "  } else if (any(findnormal) == FALSE) {\n",
    "    message(\"Series Matrix implies that this data might require normalization.\\n\")\n",
    "    isnormalized <- askYesNo(\"Is this data already normalized? \")\n",
    "    if (isnormalized == FALSE) {\n",
    "      NORM <- FALSE\n",
    "      DESEQ2DONE <- FALSE\n",
    "    }\n",
    "  }\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main Data Processing Functions\n",
    "\n",
    "Import non-GEO experiments (if no dataset already downloaded), set set data types, and perform initial processing steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "if (iscounts == FALSE) {\n",
    "  istx <- askYesNo(\"Is your dataset transcript level abundance measurements from Salmon/Kallisto/Sailfish? \")\n",
    "  if (istx == FALSE) {\n",
    "    isrsem <- askYesNo(\"Is your dataset raw abundance measurements from RSEM? \")\n",
    "    if (isrsem == TRUE) {\n",
    "      if (getgeofiles == TRUE)\n",
    "        {\n",
    "          dir <- dirname(genematrix)\n",
    "        }  #GEO\n",
    "      txtype <- 4\n",
    "      txlevel <- TRUE\n",
    "      TYPE <- \"RSEM\"\n",
    "      NORM <- FALSE\n",
    "      readline(prompt = (\"This step requires the Bioconductor package \\\"tximport\\\". Make sure it is installed, then press enter to continue...\"))\n",
    "      library(\"tximport\")\n",
    "    }\n",
    "  }\n",
    "}\n",
    "if (istx == TRUE) {\n",
    "  TYPE <- \"TXABUNDANCE\"\n",
    "  txlevel <- TRUE\n",
    "  readline(prompt = (\"This step requires the Bioconductor package \\\"tximport\\\". Make sure it is installed, then press enter to continue...\"))\n",
    "  library(\"tximport\")\n",
    "  NORM <- FALSE\n",
    "  iscounts <- FALSE\n",
    "  cat(\"\\n\")\n",
    "  cat(\"To process transcript alignments we need either an existing \\\"tx2gene\\\" file, a Gencode GTF/GFF3, or we can pull from ENSEMBL.\\n\")\n",
    "  tx2geneexists <- askYesNo(\"Do you have an existing tx2gene annotation file you wish to provide? \")\n",
    "  if (tx2geneexists == TRUE) {\n",
    "    tx2genepath <- readline(prompt = (\"Drop Your tx2gene file into R Window or Enter full path to file: \"))\n",
    "    cat(\"\\n\")\n",
    "    library(\"tools\")\n",
    "    tx2genetype <- file_ext(gsub(\".gz$\", \"\", tx2genepath))\n",
    "    if (tx2genetype == \"csv\") {\n",
    "      tx2gene <- read.table(tx2genepath, sep = \",\", header = T)\n",
    "      cat(\"Identifiers:\\n\")\n",
    "      print(as.data.frame(head(tx2gene[, 1:2], 3)))\n",
    "      txhasversions <- askYesNo(\"Do your TRANSCRIPT IDs have decimal versions? (eg. ENST00000342771.9)? \")\n",
    "      if (txhasversions == TRUE) {\n",
    "        tx2gene[, 1] <- gsub(\"\\\\..*\", \"\", tx2gene[, 1])\n",
    "        txhasversions <- FALSE\n",
    "      }\n",
    "      genehasversions <- askYesNo(\"Do your GENE IDs have decimal versions? (eg. ENSG00000158321.16)? \")\n",
    "      if (genehasversions == TRUE) {\n",
    "        tx2gene[, 2] <- gsub(\"\\\\..*\", \"\", tx2gene[, 2])\n",
    "        genehasversions <- FALSE\n",
    "      }\n",
    "    }\n",
    "    if (tx2genetype == \"txt\" | tx2genetype == \"tsv\" | tx2genetype == \"tabular\") {\n",
    "      tx2gene <- read.table(tx2genepath, sep = \"\\t\", header = T)\n",
    "      print(head(tx2gene))\n",
    "      txhasversions <- askYesNo(\"Do your TRANSCRIPT IDs have decimal versions? (eg. ENST00000342771.9)? \")\n",
    "      if (txhasversions == TRUE) {\n",
    "        tx2gene[, 1] <- gsub(\"\\\\..*\", \"\", tx2gene[, 1])\n",
    "        txhasversions <- FALSE\n",
    "      }\n",
    "      genehasversions <- askYesNo(\"Do your GENE IDs have decimal versions? (eg. ENSG00000158321.16)? \")\n",
    "      if (genehasversions == TRUE) {\n",
    "        tx2gene[, 2] <- gsub(\"\\\\..*\", \"\", tx2gene[, 2])\n",
    "        genehasversions <- FALSE\n",
    "      }\n",
    "    }\n",
    "    tx2gene <- distinct(tx2gene)\n",
    "    print(head(tx2gene))\n",
    "    cat(\"This should be formatted as a two column file with TX IDs on the left and Gene IDs on the right and no version decimals.\\n\")\n",
    "    tx2genebuild <- FALSE\n",
    "  } else if (tx2geneexists == FALSE) {\n",
    "    cat(\"Ok... \\n\")\n",
    "    tx2genebuild <- TRUE\n",
    "\n",
    "    cat(\"\\n\")\n",
    "    cat(\"We can attempt to construct the tx2gene file automatically from a GTF/GFF3 transcriptome or by pulling mappings from ENSEMBL's Biomart. \\n\")\n",
    "    cat(\"GTF/GFF3 processing requires the GenomicFeatures Package from Bioconductor to be available.\\n\")\n",
    "    cat(\"Building from ENSEMBL requires the biomaRt Package from Bioconductor to be available.\\n\")\n",
    "    cat(\"\\n\")\n",
    "    message(\"Press [1] to use mappings from a GTF/GFF3 file. Press [2] to acquire mappings from ENSEMBL. \\n\")\n",
    "    tx2genesource <- readline(prompt = (\"[1]/[2]: \"))\n",
    "    if (tx2genesource == 1) {\n",
    "\n",
    "      if (tx2genebuild == TRUE) {\n",
    "        library(\"GenomicFeatures\")\n",
    "        txgft <- readline(prompt = (\"Drop Your GTF/GFF3 file into R Window or Enter full path to file: \"))\n",
    "        TxDb <- makeTxDbFromGFF(file = txgft)\n",
    "        k <- keys(TxDb, keytype = \"TXNAME\")\n",
    "        tx2gene <- select(TxDb, k, \"GENEID\", \"TXNAME\")\n",
    "        print(head(tx2gene, 3))\n",
    "        txhasversions <- askYesNo(\"Do your TRANSCRIPT IDs have decimal versions? (eg. ENST00000342771.9)? \")\n",
    "        if (txhasversions == TRUE) {\n",
    "          tx2gene[, 1] <- gsub(\"\\\\..*\", \"\", tx2gene[, 1])\n",
    "        }\n",
    "        genehasversions <- askYesNo(\"Do your GENE IDs have decimal versions? (eg. ENSG00000158321.16)? \")\n",
    "        if (genehasversions == TRUE) {\n",
    "          tx2gene[, 2] <- gsub(\"\\\\..*\", \"\", tx2gene[, 2])\n",
    "          genehasversions <- FALSE\n",
    "        }\n",
    "        tx2geneexists <- TRUE\n",
    "        cat(\"The Tx2Gene file should be formatted as a two column file with TX IDs on the left and Gene IDs on the right.\\n\")\n",
    "        tx2gene <- distinct(tx2gene)\n",
    "      }\n",
    "      if (tx2geneexists == FALSE) {\n",
    "        cat(\"We can't continue with transcript level data without transcript to gene mappings...\\n\")\n",
    "        FAIL <- TRUE\n",
    "        if (FAIL == TRUE) {\n",
    "          stop(\"Necessary files are not available so processing was terminated.\")\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    if (tx2genesource == 2) {\n",
    "      readline(prompt = (\"This step requires the Bioconductor package \\\"biomaRt\\\". Make sure it is installed, then press enter to continue...\"))\n",
    "      library(\"biomaRt\")\n",
    "      cat(\"Default: ENSEMBL Version 97 - Human\\n\")\n",
    "      altversion <- askYesNo(\"Do you want to override this default? \")\n",
    "      if (altversion == TRUE) {\n",
    "        ensemblversion <- readline(prompt = (\"Enter the ENSEMBL version (eg. 96) you want to use (number only): \"))\n",
    "        #species <- \"hsapiens_gene_ensembl\"\n",
    "        altspecies <- askYesNo(\"The default species is HUMAN you want to override this default? \")\n",
    "        if (altspecies == TRUE){\n",
    "        datasets <- as.data.frame(listDatasets(useEnsembl(biomart=\"ensembl\",version=ensemblversion)), header =T)\n",
    "        species <- select.list(datasets[,1], title=\"Select the species of the dataset you wish to process:\")\n",
    "         } else if (altspecies == FALSE){species <- \"hsapiens_gene_ensembl\"}\n",
    "      } else if (altversion == FALSE) {\n",
    "        ensemblversion <- \"97\"\n",
    "        species <- \"hsapiens_gene_ensembl\"\n",
    "        cat(\"Using ENSEMBL97 annotations matching MSigDB7...\\n\")\n",
    "      }\n",
    "      ensmart <- useEnsembl(biomart = \"ensembl\", dataset = species,\n",
    "        version = paste0(ensemblversion))\n",
    "      cat(\"Building ENSEMBL Transcript -> Gene Symbol Mappings..\\n\")\n",
    "      rawtx2gene <- getBM(attributes = c(\"ensembl_transcript_id\", \"ensembl_gene_id\"),\n",
    "        mart = ensmart)\n",
    "      colnames(rawtx2gene) <- c(\"TXNAME\", \"GENEID\")\n",
    "      tx2gene <- distinct(rawtx2gene)\n",
    "    }\n",
    "  }\n",
    "  cat(\"\\n\")\n",
    "  message(\"How were your transcripts quantified? tximport supports the following methods:\\n\")\n",
    "  cat(\"[1] Salmon\\n\")\n",
    "  cat(\"[2] Sailfish\\n\")\n",
    "  cat(\"[3] Kallisto\\n\")\n",
    "  cat(\"[4] RSEM\\n\")\n",
    "  # cat('[5] Stringtie\\n') cat('[6] Generic Transcript Level Quantification\n",
    "  # Table\\n') cat('WARNING: ONLY SALMON [1], Sailfish [2], and KALLISTO [3] ARE\n",
    "  # CURRENTLY SUPPORTED\\n')\n",
    "  cat(\"\\n\")\n",
    "  txtype <- readline(prompt = (\"Select your platform by entering the corresponding >>NUMBER<< (without brackets): \"))\n",
    "  cat(\"\\n\")\n",
    "  if (getgeofiles == TRUE)\n",
    "    {\n",
    "      if (nrow(geoimporttable) > 1) {\n",
    "        dir <- genematrix\n",
    "      } else {\n",
    "        dir <- dirname(genematrix)\n",
    "      }\n",
    "    }  #GEO\n",
    "  if (txtype == 1) {\n",
    "    if (getgeofiles == FALSE) {\n",
    "      dir <- readline(prompt = (\"Drop a directory containing Salmon output into R Window or Enter Directory Path: \"))\n",
    "    }\n",
    "    files <- list.files(dir, recursive = TRUE, full.names = TRUE, pattern = \".sf|.sf.gz\")\n",
    "    names(files) <- paste0(tools::file_path_sans_ext(tools::file_path_sans_ext(list.files(dir,\n",
    "      recursive = TRUE, full.names = FALSE, pattern = \".sf|.sf.gz\"))))\n",
    "    if (length(files) > 0) {\n",
    "      txi.salmon <- tximport(files, type = \"salmon\", tx2gene = tx2gene, ignoreTxVersion = TRUE,\n",
    "        ignoreAfterBar = TRUE)\n",
    "      tximportcounts <- as.data.frame(txi.salmon$counts)\n",
    "    } else if (length(files) == 0) {\n",
    "      txtype <- 6\n",
    "    }\n",
    "  } else if (txtype == 2) {\n",
    "    if (getgeofiles == FALSE) {\n",
    "      dir <- readline(prompt = (\"Drop a directory containing Sailfish output into R Window or Enter Directory Path: \"))\n",
    "    }\n",
    "    files <- list.files(dir, recursive = TRUE, full.names = TRUE, pattern = \"quant.sf|quant.sf.gz\")\n",
    "    names(files) <- paste0(tools::file_path_sans_ext(tools::file_path_sans_ext(list.files(dir,\n",
    "      recursive = TRUE, full.names = FALSE, pattern = \"quant.sf|quant.sf.gz\"))))\n",
    "    if (length(files) > 0) {\n",
    "      txi.sailfish <- tximport(files, type = \"sailfish\", tx2gene = tx2gene,\n",
    "        ignoreTxVersion = TRUE, ignoreAfterBar = TRUE)\n",
    "      tximportcounts <- as.data.frame(txi.sailfish$counts)\n",
    "    } else if (length(files) == 0) {\n",
    "      txtype <- 6\n",
    "    }\n",
    "  } else if (txtype == 3) {\n",
    "    cat(\"Import Kallisto abundances from \\\"abundance.h5\\\" (requires rhdf5 library), or \\\"abundance.tsv*\\\".\\n\")\n",
    "    kallistotype <- readline(prompt = (\"Select kallisto datatype by entering \\\"h5\\\" or \\\"tsv\\\" (without quotes): \"))\n",
    "    if (kallistotype == \"h5\") {\n",
    "      readline(prompt = (\"This step requires the Bioconductor package \\\"rhdf5\\\". Make sure it is installed, then press enter to continue...\"))\n",
    "      library(\"rhdf5\")\n",
    "      if (getgeofiles == FALSE) {\n",
    "        dir <- readline(prompt = (\"Drop a directory containing Kallisto output into R Window or Enter Directory Path: \"))\n",
    "      }\n",
    "      files <- list.files(dir, recursive = TRUE, full.names = TRUE, pattern = \"abundance.h5|abundance.h5.gz\")\n",
    "      if (length(files) > 0) {\n",
    "        names(files) <- paste0(tools::file_path_sans_ext(tools::file_path_sans_ext(list.files(dir,\n",
    "          recursive = TRUE, full.names = FALSE, pattern = \"abundance.h5|abundance.h5.gz\"))))\n",
    "        txi.kallisto.h5 <- tximport(files, type = \"kallisto\", tx2gene = tx2gene,\n",
    "          ignoreTxVersion = TRUE, ignoreAfterBar = TRUE)\n",
    "        tximportcounts <- as.data.frame(txi.kallisto.h5$abundance)\n",
    "      } else if (length(files) == 0) {\n",
    "        txtype <- 6\n",
    "      }\n",
    "    }\n",
    "    if (kallistotype == \"tsv\") {\n",
    "      if (getgeofiles == FALSE) {\n",
    "        dir <- readline(prompt = (\"Drop a directory containing Kallisto output into R Window or Enter Directory Path: \"))\n",
    "      }\n",
    "      files <- list.files(dir, recursive = TRUE, full.names = TRUE, pattern = \"abundance.tsv|abundance.tsv.gz\")\n",
    "      if (length(files) > 0) {\n",
    "        names(files) <- paste0(tools::file_path_sans_ext(tools::file_path_sans_ext(list.files(dir,\n",
    "          recursive = TRUE, full.names = FALSE, pattern = \"abundance.tsv|abundance.tsv.gz\"))))\n",
    "        txi.kallisto.tsv <- tximport(files, type = \"kallisto\", tx2gene = tx2gene,\n",
    "          ignoreTxVersion = TRUE, ignoreAfterBar = TRUE)\n",
    "        tximportcounts <- as.data.frame(txi.kallisto.tsv$abundance)\n",
    "      } else if (length(files) == 0) {\n",
    "        txtype <- 6\n",
    "      }\n",
    "    }\n",
    "  } else if (txtype == 5) {\n",
    "    cat(\"Stringtie suppot not yet implemented\\n\")\n",
    "    # cat('We can import files produced by running the \\'stringtie -eB -G\n",
    "    # transcripts.gff <source_file.bam>\\' command per the TXImport tutorial.\\n')\n",
    "    # dir <- readline(prompt=('Drop a directory containing Stringtie output folder\n",
    "    # into R Window or Enter Directory Path: ')) samples <- list.files(dir) files <-\n",
    "    # file.path(dir, samples) tmp <- read_tsv(files[1]) print(head(tmp)) strtx <-\n",
    "    # readline(prompt=('Enter the column name that contains the stringtie transcript\n",
    "    # identifiers: ')) strgene <- readline(prompt=('Enter the column name that\n",
    "    # contains gene identifiers for which you have a GSEA compatible chip file (or\n",
    "    # gene symbols): ')) tx2gene <- tmp[, c(strtx, strgene)] print(head(tx2gene))\n",
    "    # readline(prompt=('If this looks correct, press enter to continue...')) txi <-\n",
    "    # tximport(tmp, type = 'stringtie', tx2gene = tx2gene)\n",
    "    FAIL <- TRUE\n",
    "  }\n",
    "  if (txtype == 6) {\n",
    "    cat(\"Failover Mode. This method isn't really supported. Use at your own risk.\\n\")\n",
    "    cat(\"Could not detect transcript quantifications with tximport.\\n\")\n",
    "    cat(\" We're going to have to make a lot of guesses here, but we'll get through this together.\\n\")\n",
    "    if (getgeofiles == FALSE) {\n",
    "      txmatrix <- readline(prompt = (\"Drop Transcript Expression Matrix into R Window or Enter File Path (supports csv or tab delimited txt): \"))\n",
    "      library(\"tools\")\n",
    "      txmatrixtype <- file_ext(gsub(\".gz$\", \"\", txmatrix))\n",
    "      if (txmatrixtype == \"csv\") {\n",
    "        full <- read.table(txmatrix, sep = \",\", header = T, row.names = NULL)\n",
    "        if (colnames(full)[1] == \"X\") {\n",
    "          colnames(full)[1] <- colnames(full[\"ID\"])\n",
    "        }\n",
    "        if (colnames(full)[1] == \"row.names\") {\n",
    "          colnames(full)[1] <- colnames(full[\"ID\"])\n",
    "        }\n",
    "      }\n",
    "      if (txmatrixtype == \"txt\" | txmatrix == \"tabular\" | txmatrix == \"tsv\" |\n",
    "        txmatrix == \"tab\") {\n",
    "        full <- read.table(txmatrix, sep = \"\\t\", header = T, row.names = NULL)\n",
    "        if (colnames(full)[1] == \"X\") {\n",
    "          colnames(full)[1] <- colnames(full[\"ID\"])\n",
    "        }\n",
    "        if (colnames(full)[1] == \"row.names\") {\n",
    "          colnames(full)[1] <- \"ID\"\n",
    "        }\n",
    "      }\n",
    "      # if(){} #Add folder parser HERE\n",
    "      importdata <- as.data.frame(colnames(full), stringsAsFactors = FALSE,\n",
    "        header = FALSE)\n",
    "      colnames(full)[1] <- \"EXPERIMENT\"\n",
    "      cat(\"\\n\")\n",
    "      displayexp <- merge(x = importdata, y = t(full[c(1:3, 11:13), ]), by.x = 1,\n",
    "        by.y = 0)\n",
    "      print(displayexp)\n",
    "      expids <- readline(prompt = (\"Enter the name from the EXPERIMENT column or row number above that defines your transcript identifiers: \"))\n",
    "      expidnumber <- match(expids, cbind(rownames(importdata), importdata)[,\n",
    "        1])\n",
    "      if (is.na(expidnumber) == TRUE) {\n",
    "        expidnumber <- match(expids, cbind(rownames(importdata), importdata)[,\n",
    "          2])\n",
    "      }\n",
    "      colnames(mergedexp)[expidnumber] <- \"TXNAME\"\n",
    "\n",
    "      cat(\"Transcript expression Matrix Imported\\n\")\n",
    "      # txlevel <- FALSE\n",
    "      cat(\"\\n\")\n",
    "      print(head(full, 3))\n",
    "      cat(\"\\n\")\n",
    "      # cat('When you're prompted for a CHIP file, instead provide a table mapping\n",
    "      # Transcript IDs to Gene Symbols and Descriptions USING CHIP HEADERS. Good\n",
    "      # Luck.\\n') readline(prompt=('Press enter to continue...'))\n",
    "    }\n",
    "    if (getgeofiles == TRUE) {\n",
    "      cat(\"Attempting to parse multiple individually quantified samples into single matrix...\\n\")\n",
    "      inputsamples <- list.files(outfile)\n",
    "      inputsamplestxt <- list.files(outfile, pattern = \".txt|.tsv|.tabular|.tab\")\n",
    "      txtsamplenumber <- length(inputsamplestxt)\n",
    "      inputsamplescsv <- list.files(outfile, pattern = \".csv\")\n",
    "      csvsamplenumber <- length(inputsamplescsv)\n",
    "      totalsamplenumber <- txtsamplenumber + csvsamplenumber\n",
    "      if (txtsamplenumber > 0) {\n",
    "        print(inputsamplestxt)\n",
    "        cat(paste(txtsamplenumber), \"tabular formatted samples were detected in input directory.\\n\")\n",
    "        txtsamplepaths <- file.path(outfile, inputsamplestxt)\n",
    "        names(txtsamplepaths) <- paste0(inputsamplestxt)\n",
    "        import_txt <- lapply(txtsamplepaths, read.table, header = FALSE,\n",
    "          row.names = NULL, sep = \"\\t\", stringsAsFactors = FALSE)\n",
    "        for (i in 1:txtsamplenumber) {\n",
    "          do\n",
    "          colnames(import_txt[[i]]) <- paste(inputsamplestxt[i], colnames(import_txt[[i]]),\n",
    "          sep = \"_\")\n",
    "        }\n",
    "      }\n",
    "      if (csvsamplenumber > 0) {\n",
    "        print(inputsamplescsv)\n",
    "        cat(paste(csvsamplenumber), \"csv formatted samples were detected in input directory.\\n\")\n",
    "        csvsamplepaths <- file.path(outfile, inputsamplescsv)\n",
    "        names(csvsamplepaths) <- paste0(inputsamplescsv)\n",
    "        import_csv <- lapply(csvsamplepaths, read.table, header = FALSE,\n",
    "          row.names = NULL, sep = \",\", stringsAsFactors = FALSE)\n",
    "        for (i in 1:csvsamplenumber) {\n",
    "          do\n",
    "          colnames(import_csv[[i]]) <- paste(inputsamplestxt[i], colnames(import_csv[[i]]),\n",
    "          sep = \"_\")\n",
    "        }\n",
    "      }\n",
    "      if ((txtsamplenumber > 0) == (csvsamplenumber > 0)) {\n",
    "        import.list <- append(import_txt, import_csv)  #combine csv and txt imports\n",
    "        cat(paste(length(import.list)), \"total samples were detected in input directory.\\n\")\n",
    "      } else if (csvsamplenumber == 0) {\n",
    "        import.list <- import_txt  #process txt only\n",
    "      } else if (txtsamplenumber == 0) {\n",
    "        import.list <- import_csv  #process csv only\n",
    "      }\n",
    "\n",
    "      keep <- askYesNo(\"Are all of the quantifications from the same pipeline? \")\n",
    "      if (keep == FALSE) {\n",
    "        list <- list(NA)\n",
    "        for (i in 1:length(names(import.list))) {\n",
    "          do\n",
    "          cat(\"\\n\")\n",
    "          cat(\"Sample: \", paste0(names(import.list[i])), \"\\n\")\n",
    "          value <- askYesNo(\"Include in analysis? \")\n",
    "          if (value == TRUE) {\n",
    "          list[[1]][i] <- names(import.list[i])\n",
    "          }\n",
    "        }\n",
    "        list <- list[[1]][!is.na(list[[1]])]\n",
    "        import.list <- import.list[list]\n",
    "      }\n",
    "\n",
    "      importdata <- as.data.frame(colnames(import.list[[1]]), stringsAsFactors = FALSE,\n",
    "        header = FALSE)\n",
    "      colnames(importdata)[1] <- \"EXPERIMENT\"\n",
    "      cat(\"\\n\")\n",
    "      displayexp <- merge(x = importdata, y = t(import.list[[1]][c(1:3, 11:13),\n",
    "        ]), by.x = 1, by.y = 0)\n",
    "      print(displayexp)\n",
    "      cat(\"\\n\")\n",
    "      cat(\"Selected a sample to use for learning dataset format: \\n\")\n",
    "      cat(\"\\n\")\n",
    "      rebuildheader <- askYesNo(\"Is there a more descriptive header than the one in the \\\"EXPERIMENT\\\" column? \")\n",
    "      if (rebuildheader == TRUE) {\n",
    "        cat(\"\\n\")\n",
    "        print(displayexp[2:4])\n",
    "        headnum <- as.numeric(readline(prompt = (\"Enter the COLUMN NUMBER above that you want to use as the labels: \")))\n",
    "        importlist2 <- lapply(import.list, function(x) {\n",
    "          colnames(x) = x[headnum, ]\n",
    "          x = x[-headnum, ]\n",
    "        })\n",
    "        import.list <- importlist2\n",
    "      }\n",
    "      expids <- readline(prompt = (\"Enter the name from the EXPERIMENT column or row number above that defines your transcript identifiers: \"))\n",
    "      expidnumber <- match(expids, cbind(rownames(importdata), importdata)[,\n",
    "        1])\n",
    "      if (is.na(expidnumber) == TRUE) {\n",
    "        expidnumber <- match(expids, cbind(rownames(importdata), importdata)[,\n",
    "          2])\n",
    "      }\n",
    "      mergedexp <- Reduce(function(x, y) merge(x, y, all = FALSE, by = expidnumber,\n",
    "        all.x = TRUE, all.y = TRUE), import.list, accumulate = F)\n",
    "      colnames(mergedexp)[expidnumber] <- \"TXNAME\"\n",
    "      mergedexp_2 <- mergedexp\n",
    "      colnames(mergedexp_2) <- make.unique(colnames(mergedexp_2))\n",
    "\n",
    "      # Disabled autocleanup - causes issues with too many formats.  mergedexp %>%\n",
    "      # select_if(is.numeric) -> mergedexp_2 mergedexp_2 <-\n",
    "      # cbind(mergedexp[expidnumber], mergedexp_2) mergedexp_2 <- mergedexp_2[,\n",
    "      # -grep('ength$', colnames(mergedexp_2))] cat('Cleaned up identifiable extraneous\n",
    "      # non-numeric columns.\\n')\n",
    "\n",
    "      fullimportdata <- as.data.frame(colnames(mergedexp_2), stringsAsFactors = FALSE,\n",
    "        header = TRUE)\n",
    "      colnames(fullimportdata)[1] <- \"EXPERIMENT\"\n",
    "      displayfullexp <- merge(x = fullimportdata, y = t(mergedexp_2[c(1:3,\n",
    "        11:13), ]), by.x = 1, by.y = 0)\n",
    "      print(displayfullexp)\n",
    "      cat(\"There are\", paste(length(colnames(mergedexp_2))), \"entries in the Experiment data matrix.\\n\")\n",
    "      keepcols <- readline(prompt = (\"How many are tx quantifications? (this should be one per sample): \"))\n",
    "      removecols <- as.numeric(length(colnames(mergedexp_2))) - (1 + as.numeric(keepcols))\n",
    "      if (removecols > 0) {\n",
    "        repeat {\n",
    "          importdataloop <- fullimportdata\n",
    "          colnames(importdataloop) <- c(\"EXPERIMENT\")\n",
    "          fullloop <- mergedexp_2\n",
    "          print(importdataloop)\n",
    "          for (i in 1:as.numeric(removecols)) {\n",
    "          do\n",
    "          cat(\"Discard everything except your Transcript IDs and your per sample quantifications. \\n\")\n",
    "          expidsdrop <- readline(prompt = (\"Enter a name or row number from the EXPERIMENT column above that defines fields you want to DISCARD: \"))\n",
    "          expiddropnumber <- match(expidsdrop, cbind(rownames(importdataloop),\n",
    "            importdataloop)[, 1])\n",
    "          if (is.na(expiddropnumber) == TRUE) {\n",
    "            expiddropnumber <- FALSE\n",
    "          }\n",
    "          if (expiddropnumber == expidsdrop) {\n",
    "            expidsdrop <- importdataloop[expiddropnumber, ]\n",
    "          }\n",
    "          if (length(expidsdrop) == \"1\") {\n",
    "            cat(\"Dropping unused identifiers\\n\")\n",
    "            cat(\"\\n\")\n",
    "            fullloop <- fullloop[, -which(names(fullloop) %in% c(expidsdrop))]\n",
    "            importdataloop <- as.data.frame(colnames(fullloop), stringsAsFactors = FALSE,\n",
    "            header = TRUE)\n",
    "            colnames(importdataloop) <- c(\"EXPERIMENT\")\n",
    "            print(importdataloop)\n",
    "            cat(\"\\n\")\n",
    "            expidsdrop <- NULL\n",
    "          }\n",
    "          }\n",
    "          check <- askYesNo(\"Does this display only a single transcript identifier and the sample ids? \")\n",
    "          if (check == TRUE) {\n",
    "          coldata <- importdataloop\n",
    "          full <- fullloop\n",
    "          expids <- expidnumber\n",
    "          print(head(tximportcounts[expidnumber], 3))\n",
    "          break\n",
    "          }\n",
    "        }\n",
    "      } else if (removecols == 0) {\n",
    "        coldata <- fullimportdata\n",
    "        full <- mergedexp_2\n",
    "        expids <- expidnumber\n",
    "        print(head(full[expidnumber], 3))\n",
    "      }\n",
    "      txhasversions <- askYesNo(\"Do your TRANCRIPT IDs have decimal versions? (eg. ENST00000342771.9)? \")\n",
    "      if (txhasversions == TRUE) {\n",
    "        full[, 1] <- gsub(\"\\\\..*\", \"\", full[, 1])\n",
    "        txhasversions <- FALSE\n",
    "      }\n",
    "\n",
    "      cat(\"\\n\")\n",
    "      cat(\"Expression Matrix Imported\\n\")\n",
    "      cat(\"\\n\")\n",
    "      print(head(full[expidnumber], 3))\n",
    "      cat(\"\\n\")\n",
    "      # cat('When you're prompted for a CHIP file, instead provide a table mapping\n",
    "      # Transcript IDs to Gene Symbols and Descriptions USING CHIP HEADERS. Good\n",
    "      # Luck.\\n') readline(prompt=('Press enter to continue...')) #Implement merge\n",
    "      # with tx2gene.\n",
    "    }\n",
    "    # Add tx2gene mapper HERE\n",
    "    full <- merge(x = tx2gene, y = full, by.x = \"TXNAME\", by.y = \"TXNAME\")\n",
    "    full <- full[, -1]\n",
    "    full <- distinct(full)\n",
    "    full <- full %>% group_by(GENEID) %>% summarise_all(sum) %>% data.frame()\n",
    "\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "if (txtype == 4) {\n",
    "  TYPE <- \"RSEM\"\n",
    "  if (getgeofiles == FALSE) {\n",
    "    dir <- readline(prompt = (\"Drop a directory containing RSEM output into R Window or Enter Directory Path: \"))\n",
    "  }\n",
    "  rsemcontents <- as.data.frame(paste0(tools::file_path_sans_ext(tools::file_path_sans_ext(list.files(dir,\n",
    "    recursive = TRUE, full.names = FALSE, pattern = \".results.gz\")))), stringsAsFactors = FALSE)\n",
    "  rsemgenes <- as.data.frame(paste0(tools::file_path_sans_ext(tools::file_path_sans_ext(list.files(dir,\n",
    "    recursive = TRUE, full.names = FALSE, pattern = \".genes.results.gz\")))),\n",
    "    stringsAsFactors = FALSE)\n",
    "  rsemtxs <- as.data.frame(paste0(tools::file_path_sans_ext(tools::file_path_sans_ext(list.files(dir,\n",
    "    recursive = TRUE, full.names = FALSE, pattern = \".isoforms.results.gz\")))),\n",
    "    stringsAsFactors = FALSE)\n",
    "  rsemgenetest <- rsemgenes[, 1] %in% rsemcontents[, 1]\n",
    "  rsemtxtest <- rsemtxs[, 1] %in% rsemcontents[, 1]\n",
    "  if (all(rsemgenetest == TRUE) == TRUE) {\n",
    "    cat(\"Gene level RSEM quantifications are available. Using these directly.\\n \")\n",
    "    USEDRSEMTXLEVEL <- FALSE\n",
    "    files <- list.files(dir, recursive = TRUE, full.names = TRUE, pattern = \".genes.results.gz\")\n",
    "    names(files) <- paste0(tools::file_path_sans_ext(tools::file_path_sans_ext(list.files(dir,\n",
    "      recursive = TRUE, full.names = FALSE, pattern = \".genes.results.gz\"))))\n",
    "    txi.rsem <- tximport(files, type = \"rsem\", txIn = FALSE, txOut = FALSE)\n",
    "    txi.rsemcounts <- as.data.frame(txi.rsem$counts)\n",
    "  } else if (all(rsemtxtest == TRUE) == TRUE) {\n",
    "    cat(\"Gene level RSEM quantifications are NOT available.\\n \")\n",
    "    cat(\"Using RSEM isoform abundances. This is a little messy.\\n \")\n",
    "    USEDRSEMTXLEVEL <- TRUE\n",
    "    files <- list.files(dir, recursive = TRUE, full.names = TRUE, pattern = \".isoforms.results.gz\")\n",
    "    names(files) <- paste0(tools::file_path_sans_ext(tools::file_path_sans_ext(list.files(dir,\n",
    "      recursive = TRUE, full.names = FALSE, pattern = \".isoforms.results.gz\"))))\n",
    "    txi.rsem <- tximport(files, type = \"rsem\", txIn = TRUE, txOut = TRUE)\n",
    "    txi.rsemcounts <- as.data.frame(txi.rsem$counts)\n",
    "  }\n",
    "  print(\"Done\")  #fixes a small bug in tximport rsem output\n",
    "  cat(\"\\n\")\n",
    "  cat(\"Identifiers:\\n\")\n",
    "  print(head(txi.rsemcounts[, 1], 3))\n",
    "  txi.rsemcounts <- tibble::rownames_to_column(txi.rsemcounts, \"GENEID\")\n",
    "  cat(\"\\n\")\n",
    "  if (all(rsemgenetest == TRUE) == TRUE) {\n",
    "    genehasversions <- askYesNo(\"Do your GENE IDs have decimal versions? (eg. ENSG00000158321.16)? \")\n",
    "  } else if (USEDRSEMTXLEVEL == TRUE) {\n",
    "    genehasversions <- askYesNo(\"Do your TRANSCRIPT IDs have decimal versions? (eg. ENST00000342771.9)? \")\n",
    "  }\n",
    "  cat(\"\\n\")\n",
    "  if (genehasversions == TRUE) {\n",
    "    txi.rsemcounts[, 1] <- gsub(\"\\\\..*\", \"\", txi.rsemcounts[, 1])\n",
    "    genehasversions <- FALSE\n",
    "  }\n",
    "  txi.rsemcounts <- distinct(txi.rsemcounts)\n",
    "  txi.rsemcounts <- txi.rsemcounts %>% group_by(GENEID) %>% summarise_all(sum) %>%\n",
    "    data.frame()\n",
    "  tximportcounts <- txi.rsemcounts\n",
    "  if (USEDRSEMTXLEVEL == TRUE) {\n",
    "    colnames(tximportcounts)[1] <- \"TXNAME\"\n",
    "    cat(\"When you're prompted for a CHIP file, instead provide a table mapping Transcript IDs to Gene Symbols and Descriptions USING CHIP HEADERS. Good Luck.\\n\")\n",
    "    message(\"THIS IS CURRENTLY BROKEN DUE TO CHIP HANDLING CHANGES, SORRY.\\n\")\n",
    "    readline(prompt = (\"Press enter to continue...\"))\n",
    "    # Add flag to redirect to separate tx2gene builder HERE or move to before\n",
    "    # txbuilder\n",
    "  }\n",
    "}\n",
    "\n",
    "if (txlevel == FALSE) {\n",
    "  if (iscounts != TRUE) {\n",
    "    iscounts <- askYesNo(\"Is your dataset gene level COUNTS measurements from HTSeq-counts, FeatureCounts, or similar? \")\n",
    "    if (iscounts == TRUE) {\n",
    "      TYPE <-  \"COUNTS\"\n",
    "      if (isnormalized == FALSE) {\n",
    "        isnormalized <- askYesNo(\"Are your counts already normalized? \")\n",
    "        if (isnormalized == TRUE) {\n",
    "          NORM <- TRUE\n",
    "          DESEQ2DONE <- FALSE\n",
    "        }\n",
    "        if (isnormalized == FALSE) {\n",
    "          NORM <- FALSE\n",
    "        }\n",
    "      } else if (isnormalized == TRUE) {\n",
    "        NORM <- TRUE\n",
    "        DESEQ2DONE <- FALSE\n",
    "      }\n",
    "    } else if (iscounts == FALSE) {\n",
    "      cat(\"Your dataset is not supported at this time.\\n\")\n",
    "      FAIL <- TRUE\n",
    "    }\n",
    "  }\n",
    "}\n",
    "\n",
    "if (FAIL == TRUE) {\n",
    "  stop(\"An unsupported datatype was encountered and processing was terminated.\")\n",
    "}\n",
    "\n",
    "# Import Gene Expression Matrix ##GEO conditional Needed for genematrix prompt\n",
    "if (txlevel == FALSE) {\n",
    "  if (getgeofiles == FALSE) {\n",
    "    genematrix <- readline(prompt = (\"Drop Gene Expression Matrix into R Window or Enter File Path (supports csv or tab delimited txt): \"))\n",
    "  }\n",
    "  library(\"tools\")\n",
    "  genematrixtype <- file_ext(gsub(\".gz$\", \"\", genematrix))\n",
    "  if (genematrixtype == \"csv\") {\n",
    "    full <- read.table(genematrix, sep = \",\", header = T, row.names = NULL)\n",
    "    if (colnames(full)[1] == \"X\") {\n",
    "      colnames(full)[1] <- \"ID\"\n",
    "    }\n",
    "    if (colnames(full)[1] == \"row.names\") {\n",
    "      colnames(full)[1] <- \"ID\"\n",
    "    }\n",
    "    cat(\"\\n\")\n",
    "    cat(\"Expression Matrix Imported\\n\")\n",
    "    cat(\"\\n\")\n",
    "    print(head(full, 3))\n",
    "    cat(\"\\n\")\n",
    "  }\n",
    "  if (genematrixtype == \"txt\" | genematrixtype == \"tabular\" | genematrixtype ==\n",
    "    \"tsv\" | genematrixtype == \"tab\") {\n",
    "    full <- read.table(genematrix, sep = \"\\t\", header = T, row.names = NULL)\n",
    "    if (colnames(full)[1] == \"X\") {\n",
    "      colnames(full)[1] <- \"ID\"\n",
    "    }\n",
    "    if (colnames(full)[1] == \"row.names\") {\n",
    "      colnames(full)[1] <- \"ID\"\n",
    "    }\n",
    "    cat(\"\\n\")\n",
    "    cat(\"Expression Matrix Imported\\n\")\n",
    "    cat(\"\\n\")\n",
    "    print(head(full, 3))\n",
    "    cat(\"\\n\")\n",
    "  }\n",
    "  if ((genematrixtype == \"\") == TRUE) {\n",
    "    is_directory <- TRUE\n",
    "    message(\"Directory detected. This directory must contain ONLY samples to be imported\\n\")\n",
    "    cat(\"Attempting to parse multiple individually quantified samples into single matrix...\\n\")\n",
    "    inputsamples <- list.files(genematrix)\n",
    "    inputsamplestxt <- list.files(genematrix, pattern = \".txt|.tsv|.tabular|.tab\")\n",
    "    txtsamplenumber <- length(inputsamplestxt)\n",
    "    inputsamplescsv <- list.files(genematrix, pattern = \".csv\")\n",
    "    csvsamplenumber <- length(inputsamplescsv)\n",
    "    totalsamplenumber <- txtsamplenumber + csvsamplenumber\n",
    "    if (txtsamplenumber > 0) {\n",
    "      print(inputsamplestxt)\n",
    "      cat(paste(txtsamplenumber), \"tabular formatted samples were detected in input directory.\\n\")\n",
    "      txtsamplepaths <- file.path(genematrix, inputsamplestxt)\n",
    "      names(txtsamplepaths) <- paste0(inputsamplestxt)\n",
    "      import_txt <- lapply(txtsamplepaths, read.table, header = FALSE, row.names = NULL,\n",
    "        sep = \"\\t\", stringsAsFactors = FALSE)\n",
    "      for (i in 1:txtsamplenumber) {\n",
    "        do\n",
    "        colnames(import_txt[[i]]) <- paste(inputsamplestxt[i], colnames(import_txt[[i]]),\n",
    "          sep = \"_\")\n",
    "      }\n",
    "    }\n",
    "    if (csvsamplenumber > 0) {\n",
    "      print(inputsamplescsv)\n",
    "      cat(paste(csvsamplenumber), \"csv formatted samples were detected in input directory.\\n\")\n",
    "      csvsamplepaths <- file.path(genematrix, inputsamplescsv)\n",
    "      names(csvsamplepaths) <- paste0(inputsamplescsv)\n",
    "      import_csv <- lapply(csvsamplepaths, read.table, header = FALSE, row.names = NULL,\n",
    "        sep = \",\", stringsAsFactors = FALSE)\n",
    "      for (i in 1:csvsamplenumber) {\n",
    "        do\n",
    "        colnames(import_csv[[i]]) <- paste(inputsamplestxt[i], colnames(import_csv[[i]]),\n",
    "          sep = \"_\")\n",
    "      }\n",
    "    }\n",
    "    if ((txtsamplenumber > 0) == (csvsamplenumber > 0)) {\n",
    "      import.list <- append(import_txt, import_csv)  #combine csv and txt imports\n",
    "      cat(paste(length(import.list)), \"total samples were detected in input directory.\\n\")\n",
    "    } else if (csvsamplenumber == 0) {\n",
    "      import.list <- import_txt  #process txt only\n",
    "    } else if (txtsamplenumber == 0) {\n",
    "      import.list <- import_csv  #process csv only\n",
    "    }\n",
    "\n",
    "    keep <- askYesNo(\"Are all of the quantifications from the same pipeline? \")\n",
    "    if (keep == FALSE) {\n",
    "      list <- list(NA)\n",
    "      for (i in 1:length(names(import.list))) {\n",
    "        do\n",
    "        cat(\"\\n\")\n",
    "        cat(\"Sample: \", paste0(names(import.list[i])), \"\\n\")\n",
    "        value <- askYesNo(\"Include in analysis? \")\n",
    "        if (value == TRUE) {\n",
    "          list[[1]][i] <- names(import.list[i])\n",
    "        }\n",
    "      }\n",
    "      list <- list[[1]][!is.na(list[[1]])]\n",
    "      import.list <- import.list[list]\n",
    "    }\n",
    "\n",
    "    importdata <- as.data.frame(colnames(import.list[[1]]), stringsAsFactors = FALSE,\n",
    "      header = FALSE)\n",
    "    colnames(importdata)[1] <- \"EXPERIMENT\"\n",
    "    cat(\"\\n\")\n",
    "    displayexp <- merge(x = importdata, y = t(import.list[[1]][c(1:3, 11:13),\n",
    "      ]), by.x = 1, by.y = 0)\n",
    "    print(displayexp)\n",
    "    cat(\"\\n\")\n",
    "    cat(\"Selected a sample to use for learning dataset format: \\n\")\n",
    "    cat(\"\\n\")\n",
    "    rebuildheader <- askYesNo(\"Is there a more descriptive header than the one in the \\\"EXPERIMENT\\\" column? \")\n",
    "    if (rebuildheader == TRUE) {\n",
    "      cat(\"\\n\")\n",
    "      print(displayexp[2:4])\n",
    "      headnum <- as.numeric(readline(prompt = (\"Enter the COLUMN NUMBER above that you want to use as the descriptors: \")))\n",
    "      importlist2 <- lapply(import.list, function(x) {\n",
    "        colnames(x) = x[headnum, ]\n",
    "        x = x[-headnum, ]\n",
    "      })\n",
    "      import.list <- importlist2\n",
    "    }\n",
    "    expids <- readline(prompt = (\"Enter the name from the EXPERIMENT column or row number above that defines your gene identifiers: \"))\n",
    "    expidnumber <- match(expids, cbind(rownames(importdata), importdata)[, 1])\n",
    "    if (is.na(expidnumber) == TRUE) {\n",
    "      expidnumber <- match(expids, cbind(rownames(importdata), importdata)[,\n",
    "        2])\n",
    "    }\n",
    "    mergedexp <- Reduce(function(x, y) merge(x, y, all = FALSE, by = expidnumber,\n",
    "      all.x = TRUE, all.y = TRUE), import.list, accumulate = F)\n",
    "    colnames(mergedexp)[expidnumber] <- \"GENEID\"\n",
    "    mergedexp_2 <- mergedexp\n",
    "    colnames(mergedexp_2) <- make.unique(colnames(mergedexp_2))\n",
    "\n",
    "    # Disabled autocleanup - causes issues with too many formats.  mergedexp %>%\n",
    "    # select_if(is.numeric) -> mergedexp_2 mergedexp_2 <-\n",
    "    # cbind(mergedexp[expidnumber], mergedexp_2) mergedexp_2 <- mergedexp_2[,\n",
    "    # -grep('ength$', colnames(mergedexp_2))] cat('Cleaned up identifiable extraneous\n",
    "    # non-numeric columns.\\n')\n",
    "\n",
    "    fullimportdata <- as.data.frame(colnames(mergedexp_2), stringsAsFactors = FALSE,\n",
    "      header = TRUE)\n",
    "    colnames(fullimportdata)[1] <- \"EXPERIMENT\"\n",
    "    displayfullexp <- merge(x = fullimportdata, y = t(mergedexp_2[c(1:3, 11:13),\n",
    "      ]), by.x = 1, by.y = 0)\n",
    "    print(displayfullexp)\n",
    "    cat(\"There are\", paste(length(colnames(mergedexp_2))), \"entries in the Experiment data matrix.\\n\")\n",
    "    keepcols <- readline(prompt = (\"How many are gene quantifications? (this should be one per sample): \"))\n",
    "    removecols <- as.numeric(length(colnames(mergedexp_2))) - (1 + as.numeric(keepcols))\n",
    "    if (removecols > 0) {\n",
    "      repeat {\n",
    "        importdataloop <- fullimportdata\n",
    "        colnames(importdataloop) <- c(\"EXPERIMENT\")\n",
    "        fullloop <- mergedexp_2\n",
    "        print(importdataloop)\n",
    "        for (i in 1:as.numeric(removecols)) {\n",
    "          do\n",
    "          message(\"Discard everything except your Gene IDs and your per sample quantifications. \\n\")\n",
    "          expidsdrop <- readline(prompt = (\"Enter a name or row number from the EXPERIMENT column above that defines fields you want to DISCARD: \"))\n",
    "          expiddropnumber <- match(expidsdrop, cbind(rownames(importdataloop),\n",
    "          importdataloop)[, 1])\n",
    "          if (is.na(expiddropnumber) == TRUE) {\n",
    "          expiddropnumber <- FALSE\n",
    "          }\n",
    "          if (expiddropnumber == expidsdrop) {\n",
    "          expidsdrop <- importdataloop[expiddropnumber, ]\n",
    "          }\n",
    "          if (length(expidsdrop) == \"1\") {\n",
    "          cat(\"Dropping unused identifiers\\n\")\n",
    "          cat(\"\\n\")\n",
    "          fullloop <- fullloop[, -which(names(fullloop) %in% c(expidsdrop))]\n",
    "          importdataloop <- as.data.frame(colnames(fullloop), stringsAsFactors = FALSE,\n",
    "            header = TRUE)\n",
    "          colnames(importdataloop) <- c(\"EXPERIMENT\")\n",
    "          print(importdataloop)\n",
    "          cat(\"\\n\")\n",
    "          expidsdrop <- NULL\n",
    "          }\n",
    "        }\n",
    "        check <- askYesNo(\"Does this display only a single gene identifier and the sample ids? \")\n",
    "        if (check == TRUE) {\n",
    "          coldata <- importdataloop\n",
    "          full2 <- fullloop\n",
    "          expids <- expidnumber\n",
    "          displayframe <- as.data.frame(full2[expidnumber][c(1:3, 11:13),\n",
    "          ])\n",
    "          colnames(displayframe) <- c(\"GENEID\")\n",
    "          print(displayframe)\n",
    "          break\n",
    "        }\n",
    "      }\n",
    "    } else if (removecols == 0) {\n",
    "      coldata <- fullimportdata\n",
    "      full2 <- mergedexp_2\n",
    "      expids <- expidnumber\n",
    "      displayframe <- as.data.frame(full2[expidnumber][c(1:3, 11:13), ])\n",
    "      colnames(displayframe) <- c(\"GENEID\")\n",
    "      print(displayframe)\n",
    "    }\n",
    "    genehasversions <- askYesNo(\"Do your GENE IDs have decimal versions? (eg. ENSG00000158321.16)? \")\n",
    "    if (genehasversions == TRUE) {\n",
    "      full2[, 1] <- gsub(\"\\\\..*\", \"\", full2[, 1])\n",
    "      genehasversions <- FALSE\n",
    "    }\n",
    "\n",
    "    cat(\"\\n\")\n",
    "    cat(\"Expression Matrix Imported\\n\")\n",
    "    cat(\"\\n\")\n",
    "    displayframe <- as.data.frame(full2[expidnumber][c(1:3, 11:13), ])\n",
    "    colnames(displayframe) <- c(\"GENEID\")\n",
    "    print(displayframe)\n",
    "    cat(\"\\n\")\n",
    "    expids <- colnames(full2)[expids]\n",
    "  }\n",
    "\n",
    "} else if (txlevel == TRUE) {\n",
    "  full <- tximportcounts\n",
    "  if (txtype != 6) {\n",
    "    cat(\"Using TXImport result...\\n\")\n",
    "  }\n",
    "  coldata <- as.data.frame(colnames(full), stringsAsFactors = FALSE, header = FALSE)\n",
    "  coldata <- rbind(c(\"GENEID\"), coldata)\n",
    "  colnames(coldata) <- c(\"EXPERIMENT\")\n",
    "  expids <- 0\n",
    "  full2 <- full\n",
    "  cat(\"\\n\")\n",
    "  print(head(full2, 3))\n",
    "  cat(\"\\n\")\n",
    "}\n",
    "\n",
    "if ((txlevel == FALSE | TYPE == \"RSEM\") == (is_directory == FALSE)) {\n",
    "  cat(\"There are\", paste(length(colnames(full))), \"columns in your data table.\\n\")\n",
    "  samplesize <- readline(prompt = (\"How many of these are sequenced SAMPLES? \"))\n",
    "  geneids <- (length(colnames(full)) - as.numeric(samplesize))\n",
    "  repeat {\n",
    "    if (geneids == \"1\") {\n",
    "      # Prompt User for Original Experiment Namespace to Merge with CHIP On\n",
    "      message(\"Gene Expression File Header:\\n\")\n",
    "      cat(\"\\n\")\n",
    "      coldata <- as.data.frame(colnames(full), stringsAsFactors = FALSE, header = FALSE)\n",
    "      colnames(coldata) <- c(\"EXPERIMENT\")\n",
    "      print(as.data.frame(coldata))\n",
    "      cat(\"\\n\")\n",
    "      expids <- readline(prompt = (\"Enter the name from the EXPERIMENT column or row number above that defines your gene identifiers: \"))\n",
    "      expidnumber <- match(expids, cbind(rownames(coldata), coldata)[, 1])\n",
    "      if (is.na(expidnumber) == TRUE) {\n",
    "        expidnumber <- FALSE\n",
    "      }\n",
    "      if (expidnumber == expids) {\n",
    "        expids <- coldata[expidnumber, ]\n",
    "      }\n",
    "      # FIX ENSG_SYMBOL MERGE\n",
    "      if (txlevel == FALSE) {\n",
    "        if (all((grepl(\"_\", full[, expids], fixed = TRUE))) == TRUE) {\n",
    "          readline(prompt = (\"This step requires the CRAN package \\\"tidyr\\\". Make sure it is installed, then press enter to continue...\"))\n",
    "          library(\"tidyr\")\n",
    "          full <- separate(full, expids, c(\"GENEID\", NA), sep = \"_\", remove = TRUE,\n",
    "          extra = \"warn\")\n",
    "          expids <- \"GENEID\"\n",
    "        }\n",
    "      }\n",
    "      full2 <- full\n",
    "      cat(\"Gene IDs:\\n\")\n",
    "      print(head(full2[, expids], 3))\n",
    "      if (txlevel == FALSE) {\n",
    "        genehasversions <- askYesNo(\"Do your GENE IDs have decimal versions? (eg. ENSG00000158321.16)? \")\n",
    "        if (genehasversions == TRUE) {\n",
    "          full2[, 1] <- gsub(\"\\\\..*\", \"\", full2[, 1])\n",
    "          genehasversions <- FALSE\n",
    "        }\n",
    "      }\n",
    "      coldata <- as.data.frame(colnames(full2), stringsAsFactors = FALSE, header = TRUE)\n",
    "      colnames(coldata) <- c(\"EXPERIMENT\")\n",
    "      break\n",
    "    } else if (geneids > \"1\") {\n",
    "      fullloop <- full\n",
    "      # Prompt User for Original Experiment Namespace to Merge with CHIP On\n",
    "      cat(\"Gene Expression File Header:\\n\")\n",
    "      cat(\"\\n\")\n",
    "      coldata <- as.data.frame(colnames(full), stringsAsFactors = FALSE, header = FALSE)\n",
    "      colnames(coldata) <- c(\"EXPERIMENT\")\n",
    "      print(as.data.frame(coldata))\n",
    "      cat(\"\\n\")\n",
    "      message(\"We now need to pick one set of identifiers to use for downstream processing, and prune away the others one at a time.\\n\")\n",
    "      message(\"For example, if you have both ENSEMBL Gene IDs and Gene Symbols columns KEEP the ENSEMBL IDs and DISCARD the Gene Symbols\\n\")\n",
    "      cat(\"\\n\")\n",
    "      expids <- readline(prompt = (\"Enter the name from the EXPERIMENT column or row number above that defines the gene identifiers you want to KEEP: \"))\n",
    "      expidnumber <- match(expids, cbind(rownames(coldata), coldata)[, 1])\n",
    "      if (is.na(expidnumber) == TRUE) {\n",
    "        expidnumber <- FALSE\n",
    "      }\n",
    "      if (expidnumber == expids) {\n",
    "        expids <- coldata[expidnumber, ]\n",
    "      }\n",
    "      coldataloop <- coldata\n",
    "      for (i in 1:(as.numeric(geneids) - 1)) {\n",
    "        do\n",
    "        expidsdrop <- readline(prompt = (\"Enter the name from the EXPERIMENT column or row number above that defines gene identifiers you want to DISCARD: \"))\n",
    "        expiddropnumber <- match(expidsdrop, cbind(rownames(coldataloop),\n",
    "          coldataloop)[, 1])\n",
    "        if (is.na(expiddropnumber) == TRUE) {\n",
    "          expiddropnumber <- FALSE\n",
    "        }\n",
    "        if (expiddropnumber == expidsdrop) {\n",
    "          expidsdrop <- coldataloop[expiddropnumber, ]\n",
    "        }\n",
    "        if (length(expidsdrop) == \"1\") {\n",
    "          cat(\"Dropping unused identifiers\\n\")\n",
    "          cat(\"\\n\")\n",
    "          fullloop <- fullloop[, -which(names(fullloop) %in% c(expidsdrop))]\n",
    "          coldataloop <- as.data.frame(colnames(fullloop), stringsAsFactors = FALSE,\n",
    "          header = TRUE)\n",
    "          colnames(coldataloop) <- c(\"EXPERIMENT\")\n",
    "          print(coldataloop)\n",
    "          cat(\"\\n\")\n",
    "          expidsdrop <- NULL\n",
    "        }\n",
    "      }\n",
    "      # check <- askYesNo('Does this look right?')} if(check == TRUE){ }\n",
    "    }\n",
    "    check <- askYesNo(\"Does this display only a single gene identifier and the sample ids? \")\n",
    "    if (check == TRUE) {\n",
    "      coldata <- coldataloop\n",
    "      full2 <- fullloop\n",
    "      print(head(full2, 3))\n",
    "      if (TYPE == \"COUNTS\" | TYPE == \"TXABUNDANCE\") {\n",
    "        genehasversions <- askYesNo(\"Do your GENE IDs have decimal versions? (eg. ENSG00000158321.16)? \")\n",
    "      } else if (TYPE == \"RSEM\") {\n",
    "        genehasversions <- FALSE\n",
    "      }\n",
    "      if (genehasversions == TRUE) {\n",
    "        full2[, 1] <- gsub(\"\\\\..*\", \"\", full2[, 1])\n",
    "        genehasversions <- FALSE\n",
    "      }\n",
    "      break\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Catch and rename long sample IDs (optional)\n",
    "\n",
    "Some imported datasets result in extremely long sample names resulting from messy importing of multiple data files, this allows an easy check to correct this."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (median(nchar(colnames(full2))) > 25) {\n",
    "  cat(\"Your sample names are pretty long.\\n\")\n",
    "  replacenames <- askYesNo(\"Would you like to replace them with something friendlier? \")\n",
    "  if (replacenames == TRUE) {\n",
    "    repeat {\n",
    "      fullrename <- full2\n",
    "      replacenameslength <- length(colnames(fullrename))\n",
    "      for (i in 1:(replacenameslength)) {\n",
    "        do\n",
    "        message(paste0(\"[\", i, \"] \", colnames(fullrename[i])))\n",
    "        newname <- readline(prompt = (\"What would you like to call this sample? \"))\n",
    "        colnames(fullrename)[i] <- newname\n",
    "      }\n",
    "      coldatarename <- as.data.frame(colnames(fullrename), stringsAsFactors = FALSE,\n",
    "        header = TRUE)\n",
    "      colnames(coldatarename) <- c(\"EXPERIMENT\")\n",
    "      cat(\"\\n\")\n",
    "      cat(\"Formatted experiment:\\n\")\n",
    "      print(coldatarename)\n",
    "      cat(\"\\n\")\n",
    "      check3 <- askYesNo(\"Do the new sample names appear correct (make absolutely sure before continuing)? \")\n",
    "      if (check3 == TRUE) {\n",
    "        coldata <- coldatarename\n",
    "        full2 <- fullrename\n",
    "        break\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sample to Phenotype Assignment\n",
    "This function allows you to annotate your dataset to construct the files necessicary to write a GSEA compatible CLS file, as well as the coldata file required to normalize with DESeq2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "message(\"Now we'll build the GSEA CLS file using a DESeq2 \\\"coldata\\\" file.\\n\")\n",
    "colavailable <- askYesNo(\"Do you have an existing coldata file? (If \\\"no\\\", we'll build one next.) \")\n",
    "if (colavailable == TRUE) {\n",
    "  design <- readline(prompt = (\"Drop DESeq2 formatted coldata File into R Window or Enter File Path, see DESeq2 Tutorial for Formatting (https://bioconductor.org/packages/3.8/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#input-data): \"))\n",
    "  coldata <- read.table(design, sep = \"\\t\", header = T, row.names = 1)\n",
    "  colnames(coldata) <- c(\"condition\")\n",
    "} else if (colavailable == FALSE) {\n",
    "  cat(\"Okay. We'll build annotations from scratch...\\n\")\n",
    "  # construct coldata directly\n",
    "  coldata <- cbind(coldata, c(\"\"), stringsAsFactors = FALSE)\n",
    "  if (expids != 0) {\n",
    "    idrow <- which(grepl(expids, coldata))\n",
    "    coldata <- as.data.frame(coldata[-c(idrow), ], stringsAsFactors = FALSE)\n",
    "  }\n",
    "  rownames(coldata) <- 1:nrow(coldata)\n",
    "  colnames(coldata) <- c(\"name\", \"condition\")\n",
    "  looplen <- length(rownames(coldata))\n",
    "  # levels(coldata$condition)= c('EX', 'CT', 'ex', 'ct', 'Experimental', 'Control',\n",
    "  # 'experimental', 'control')\n",
    "  cat(\"\\n\")\n",
    "  cat(\"We now need to generate an experimental labels file to build the GSEA CLS file.\\n\")\n",
    "  cat(\"\\n\")\n",
    "  repeat {\n",
    "    phnumber <- readline(prompt = (\"How many phenotypes are in your experiment? (numbers only) \"))\n",
    "    phenotypes <- 0\n",
    "    for (i in 1:phnumber) {\n",
    "      do\n",
    "      cat(\"What do you want to name phenotype\", paste0(i), \"? \")\n",
    "      phenotypes[i] <- readline()\n",
    "    }\n",
    "    cat(\"\\n\")\n",
    "    print(as.data.frame(phenotypes))\n",
    "    cat(\"\\n\")\n",
    "    Phenotypeframe <- as.data.frame(phenotypes, stringsAsFactors = FALSE)\n",
    "    for (i in 1:looplen) {\n",
    "      do\n",
    "      cat(\"Select a phenotype for sample >\", paste(coldata[i, 1]), \": \")\n",
    "      value <- readline()\n",
    "      phenotypesnumber <- match(value, cbind(rownames(Phenotypeframe), Phenotypeframe)[,\n",
    "        1])\n",
    "      if (is.na(phenotypesnumber) == TRUE) {\n",
    "        coldata[i, 2] <- value\n",
    "      } else if (phenotypesnumber == value) {\n",
    "        coldata[i, 2] <- Phenotypeframe[value, ]\n",
    "      }\n",
    "    }\n",
    "\n",
    "    cat(\"\\n\")\n",
    "    print(as.data.frame(coldata))\n",
    "    cat(\"\\n\")\n",
    "    cat(\"Make absolutely sure the conditions are listed correctly before continuing!\\n\")\n",
    "    cat(\"We detected\", paste(length(unique(coldata$condition))), \"phenotypes:\",\n",
    "      paste(unique(coldata$condition)), \"\\n\")\n",
    "    check2 <- askYesNo(\"Does the sample to phenotype assignment appear correct (make absolutely sure before continuing)? \")\n",
    "    if (check2 == TRUE) {\n",
    "      rownames(coldata) <- coldata$name\n",
    "      coldata <- coldata[c(2)]\n",
    "      break\n",
    "    }\n",
    "  }\n",
    "  if (NORM == FALSE) {\n",
    "    seondaryfactor <- askYesNo(\"Do you want to annotate a secondary factor (eg: sex or batch) to include in DESeq2 normalization? \")\n",
    "  }\n",
    "  if (seondaryfactor == TRUE) {\n",
    "    repeat {\n",
    "      factorlevel <- readline(prompt = (\"What do you want to call this factor? \"))\n",
    "      newfactor <- as.data.frame(1:nrow(coldata), stringsAsFactors = FALSE)\n",
    "      colnames(newfactor) <- c(factorlevel)\n",
    "      factornumber <- readline(prompt = (\"How many levels are in this factor? (numbers only) \"))\n",
    "      factortypes <- 0\n",
    "      for (i in 1:factornumber) {\n",
    "        do\n",
    "        cat(\"What do you want to name factor level\", paste0(i), \"? \")\n",
    "        factortypes[i] <- readline()\n",
    "      }\n",
    "      cat(\"\\n\")\n",
    "      print(as.data.frame(factortypes))\n",
    "      cat(\"\\n\")\n",
    "      factorframe <- as.data.frame(factortypes, stringsAsFactors = FALSE)\n",
    "      for (i in 1:looplen) {\n",
    "        do\n",
    "        cat(\"Select a factor level for sample >\", paste(rownames(coldata)[i]),\n",
    "          \": \")\n",
    "        value <- readline()\n",
    "        factortypesnumber <- match(value, cbind(rownames(factorframe), factorframe)[,\n",
    "          1])\n",
    "        if (is.na(factortypesnumber) == TRUE) {\n",
    "          newfactor[i, 1] <- value\n",
    "        } else if (factortypesnumber == value) {\n",
    "          newfactor[i, 1] <- factorframe[value, ]\n",
    "        }\n",
    "      }\n",
    "      coldatavalidate <- cbind(coldata, newfactor, stringsAsFactors = FALSE)\n",
    "      cat(\"\\n\")\n",
    "      print(as.data.frame(coldatavalidate))\n",
    "      cat(\"\\n\")\n",
    "      cat(\"Make absolutely sure the conditions are listed correctly before continuing!\\n\")\n",
    "      cat(\"We detected\", paste(length(unique(newfactor[, 1]))), \"factor levels:\",\n",
    "        paste(unique(newfactor[, 1])), \"\\n\")\n",
    "      check2 <- askYesNo(\"Does the sample to factor assignment appear correct (make absolutely sure before continuing)? \")\n",
    "      if (check2 == TRUE) {\n",
    "        coldata <- coldatavalidate\n",
    "        break\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Perform DESeq2 data normalization (optional)\n",
    "Apply a basic low count filter to remove non-expressed genes from the matrix then apply DESeq2 normalization for statistically valid between-sample comparisons (default: normalization by size factors)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if (NORM == FALSE) {\n",
    "  cat(\"Your dataset is flagged as needing normalization to be compatible with GSEA.\\n\")\n",
    "  donormalize <- askYesNo(\"We can now normalize your dataset with DESEq2. Continue? \")\n",
    "  if (donormalize == FALSE) {\n",
    "    cat(\"Perofrming standard filtering of low count genes without additional normalization.\\n\")\n",
    "    cat(\"You probably don't want to do this. We don't think this dataset is properly normalized.\\n\")\n",
    "    NORM <- TRUE\n",
    "  }\n",
    "  if (donormalize == TRUE) {\n",
    "\n",
    "    # SUM Counts for Identifiers Mapping to the Same Gene\n",
    "    if (expids != 0) {\n",
    "      full2 <- distinct(full2)\n",
    "      full2_sum <- full2 %>% group_by(.dots = expids) %>% summarise_all(sum) %>% data.frame()\n",
    "\n",
    "      # Set Gene Names as Index Column\n",
    "      full2_sum2 <- full2_sum[, -1]\n",
    "      rownames(full2_sum2) <- full2_sum[, 1]\n",
    "      rownames(coldata) <- colnames(full2_sum2)\n",
    "      full2 <- full2_sum2\n",
    "    }\n",
    "    full3 <- full2\n",
    "#    cat(\"\\n\")\n",
    "#    outprefix <- readline(prompt = (\"Enter a prefix to label output files: \"))\n",
    "    cat(\"\\n\")\n",
    "\n",
    "    readline(prompt = (\"This step requires the Bioconductor package \\\"DESeq2\\\". Make sure it is installed, then press enter to continue...\"))\n",
    "    cat(\"Loading DESeq2 Library...\\n\")\n",
    "    library(\"DESeq2\")\n",
    "    cat(\"Begin DESeq2 Normalization...\\n\")\n",
    "    full3 <- round(full3)\n",
    "if (seondaryfactor == TRUE) {\n",
    "des <- formula(paste(\"~\",paste(colnames(coldata)[2], colnames(coldata)[1], sep=\"+\")))\n",
    "dds <- DESeqDataSetFromMatrix(countData = full3, colData = coldata, design = des)\n",
    "}\n",
    "else if (seondaryfactor == FALSE) {\n",
    "dds <- DESeqDataSetFromMatrix(countData = full3, colData = coldata, design = ~ condition)\n",
    "}\n",
    "    keep <- rowSums(counts(dds)) >= 10\n",
    "    dds <- dds[keep, ]\n",
    "    dds <- DESeq(dds)\n",
    "    res <- results(dds)\n",
    "    DESEQ2DONE <- TRUE\n",
    "  }\n",
    "}\n",
    "\n",
    "## Move Normalization functions to before chip processing, but leave merge with\n",
    "## chip after. Needs Conditional merge with chip function.\n",
    "if (DESEQ2DONE == TRUE) {\n",
    "  cat(\"Normalizing by size factors (default)\\n\")\n",
    "  dds <- estimateSizeFactors(dds)\n",
    "  norm <- counts(dds, normalized = TRUE)\n",
    "  norm <- tibble::rownames_to_column(as.data.frame(norm), \"NAME\")\n",
    "}\n",
    "\n",
    "if (altnorm == \"usevst\" | altnorm == \"useall\") {\n",
    "  cat(\"Normalizing using the DESeq2 variance stabilizing transformation\\n\")\n",
    "  vsd <- vst(dds, blind = FALSE)\n",
    "  vstnorm <- as.data.frame(assay(vsd))\n",
    "  vstnorm <- tibble::rownames_to_column(as.data.frame(vstnorm), \"NAME\")\n",
    "}\n",
    "\n",
    "if (altnorm == \"userlog\" | altnorm == \"useall\") {\n",
    "  cat(\"Normalizing using the DESeq2 rlog transformation\\n\")\n",
    "  rld <- rlog(dds, blind = FALSE)\n",
    "  rlognorm <- as.data.frame(assay(rld))\n",
    "  rlognorm <- tibble::rownames_to_column(as.data.frame(rlognorm), \"NAME\")\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Construct or import an MSigDB CHIP file\n",
    "This step will either construct, or use a provided CHIP file to convert processed datasets into the correct namespace for analysis using MSigDB gene sets. Can alternately support the conversion to any appropriate namespace using user-supplied CHIP files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import CHIP File for Processing\n",
    "message(\"We now need to convert your gene identifiers into the MSigDB namespace using GSEA CHIP files.\\n\")\n",
    "message(\"If your experiment uses >> HUMAN ENSEMBL IDs << we can do this automatically\\n\")\n",
    "buildchip <- askYesNo(\"Do you want to build the CHIP automatically? \")\n",
    "if (buildchip == TRUE) {\n",
    "  readline(prompt = (\"This requires the Bioconductor package \\\"biomaRt\\\". Make sure it is installed, then press enter to continue...\"))\n",
    "  library(\"biomaRt\")\n",
    "  cat(\"MsigDB7 uses annotations from ENSEMBL97\\n\")\n",
    "  altversion <- askYesNo(\"Do you want to override this selection? \")\n",
    "  if (altversion == TRUE) {\n",
    "    ensemblversion <- readline(prompt = (\"Enter the ENSEMBL version (eg. 96) you want to use (number only): \"))\n",
    "  } else if (altversion == FALSE) {\n",
    "    ensemblversion <- \"97\"\n",
    "    cat(\"Using ENSEMBL97 annotations matching MSigDB7...\\n\")\n",
    "  }\n",
    "  ensmart <- useEnsembl(biomart = \"ensembl\", dataset = \"hsapiens_gene_ensembl\",\n",
    "    version = paste0(ensemblversion))\n",
    "\n",
    "  cat(\"Building ENSEMBL Gene ID -> Gene Symbol Mappings..\\n\")\n",
    "  rawchip <- getBM(attributes = c(\"ensembl_gene_id\", \"external_gene_name\",\n",
    "    \"description\"), mart = ensmart)\n",
    "  colnames(rawchip) <- c(\"Probe.Set.ID\", \"Gene.Symbol\", \"Gene.Title\")\n",
    "  rawchip <- distinct(rawchip)\n",
    "\n",
    "}\n",
    "if (buildchip == FALSE) {\n",
    "  CHIPpath <- readline(prompt = (\"Drop appropriate CHIP File matching the namespace of identifiers into R Window or Enter File Path: \"))\n",
    "  rawchip <- read.table(CHIPpath, sep = \"\\t\", comment.char = \"\", quote = \"\", stringsAsFactors = FALSE,\n",
    "    fill = TRUE, header = T)\n",
    "}\n",
    "chip <- rawchip[c(\"Probe.Set.ID\", \"Gene.Symbol\")]\n",
    "colnames(chip) <- c(\"Raw_IDs\", \"NAME\")\n",
    "fullchip <- unique(rawchip[c(\"Gene.Symbol\", \"Gene.Title\")])\n",
    "colnames(fullchip) <- c(\"NAME\", \"Description\")\n",
    "\n",
    "cat(\"Done\\n\")\n",
    "cat(\"\\n\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Apply CHIP file and write out processed results\n",
    "Final data processing to construct GSEA compatible output files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " Merge Gene Expression Matrix with CHIP File and Process Identifiers\n",
    "mappedexp <- merge(x = chip, y = full2, by.x = \"Raw_IDs\", by.y = expids, all = FALSE)\n",
    "size <- length(colnames(mappedexp))\n",
    "mappedexp <- mappedexp[c(2:size)]\n",
    "\n",
    "cat(\"Summing Counts that mapped to the same gene after applying mapping chip\\n\")\n",
    "\n",
    "# SUM Counts for Identifiers Mapping to the Same Gene\n",
    "mappedexp <- distinct(mappedexp)\n",
    "mappedexp_sum <- mappedexp %>% group_by(NAME) %>% summarise_all(sum) %>% data.frame()\n",
    "\n",
    "# Set Gene Names as Index Column\n",
    "mappedexp_sum2 <- mappedexp_sum[, -1]\n",
    "rownames(mappedexp_sum2) <- mappedexp_sum[, 1]\n",
    "rownames(coldata) <- colnames(mappedexp_sum2)\n",
    "cat(\"Done\\n\")\n",
    "outprefix <- readline(prompt = (\"Enter a prefix to label output files: \"))\n",
    "cat(\"\\n\")\n",
    "\n",
    "\n",
    "if (DESEQ2DONE == TRUE) {\n",
    "  cat(\"Writing Size Factor Normalizated GCT\\n\")\n",
    "  protoGCT <- merge(x = fullchip, y = mappedexp_sum2, by.x = \"NAME\", by.y = 0, all.y = TRUE)\n",
    "  bound <- rbind(colnames(protoGCT), protoGCT)\n",
    "  bound <- rbind(NA, bound)\n",
    "  bound <- rbind(NA, bound)\n",
    "  bound[1, 1] <- \"#1.2\"\n",
    "  numberofsamples <- length(colnames(bound)) - 2\n",
    "  numberofgenes <- length(bound$NAME) - 3\n",
    "  bound[2, 1] <- numberofgenes\n",
    "  bound[2, 2] <- numberofsamples\n",
    "  write.table(bound, paste0(outprefix, \"_Default_Normalized_Counts.gct\"), sep = \"\\t\",\n",
    "    quote = F, row.names = FALSE, col.names = FALSE, na = \"\")\n",
    "}\n",
    "\n",
    "if (NORM == TRUE) {\n",
    "  cat(\"You've indicated that you dataset is already normalized.\\n\")\n",
    "  cat(\"Perofrming standard filtering of low count genes without additional normalization.\\n\")\n",
    "  mappedexp_sum3 <- subset(mappedexp_sum2, rowSums(mappedexp_sum2[]) >= 5)\n",
    "  protoGCT <- merge(x = fullchip, y = mappedexp_sum3, by.x = \"NAME\", by.y = 0,\n",
    "    all = FALSE)\n",
    "  bound <- rbind(colnames(protoGCT), protoGCT)\n",
    "  bound <- rbind(NA, bound)\n",
    "  bound <- rbind(NA, bound)\n",
    "  bound[1, 1] <- \"#1.2\"\n",
    "  numberofsamples <- length(colnames(bound)) - 2\n",
    "  numberofgenes <- length(bound$NAME) - 3\n",
    "  bound[2, 1] <- numberofgenes\n",
    "  bound[2, 2] <- numberofsamples\n",
    "  cat(\"Writing final .GCT file for GSEA\\n\")\n",
    "  write.table(bound, paste0(outprefix, \"_Formatted.gct\"), sep = \"\\t\", quote = F,\n",
    "    row.names = FALSE, col.names = FALSE, na = \"\")\n",
    "}\n",
    "\n",
    "if (altnorm == \"usevst\" | altnorm == \"useall\") {\n",
    "  cat(\"Writing Variance Stabilizing Transformation Normalizated GCT\\n\")\n",
    "  vstprotoGCT <- merge(x = fullchip, y = vstnorm, by.x = \"NAME\", by.y = 0,\n",
    "    all.y = TRUE)\n",
    "  vstbound <- rbind(colnames(vstprotoGCT), vstprotoGCT)\n",
    "  vstbound <- rbind(NA, vstbound)\n",
    "  vstbound <- rbind(NA, vstbound)\n",
    "  vstbound[1, 1] <- \"#1.2\"\n",
    "  vstnumberofsamples <- length(colnames(vstbound)) - 2\n",
    "  vstnumberofgenes <- length(vstbound$NAME) - 3\n",
    "  vstbound[2, 1] <- vstnumberofgenes\n",
    "  vstbound[2, 2] <- vstnumberofsamples\n",
    "  write.table(vstbound, paste0(outprefix, \"_vst_normalized_Counts.gct\"), sep = \"\\t\",\n",
    "    quote = F, row.names = FALSE, col.names = FALSE, na = \"\")\n",
    "}\n",
    "\n",
    "if (altnorm == \"userlog\" | altnorm == \"useall\") {\n",
    "  cat(\"Writing RLOG Normalizated GCT\\n\")\n",
    "  rlogprotoGCT <- merge(x = fullchip, y = rlognorm, by.x = \"NAME\", by.y = 0,\n",
    "    all.y = TRUE)\n",
    "  rlogbound <- rbind(colnames(rlogprotoGCT), rlogprotoGCT)\n",
    "  rlogbound <- rbind(NA, rlogbound)\n",
    "  rlogbound <- rbind(NA, rlogbound)\n",
    "  rlogbound[1, 1] <- \"#1.2\"\n",
    "  rlognumberofsamples <- length(colnames(rlogbound)) - 2\n",
    "  rlognumberofgenes <- length(rlogbound$NAME) - 3\n",
    "  rlogbound[2, 1] <- rlognumberofgenes\n",
    "  rlogbound[2, 2] <- rlognumberofsamples\n",
    "  write.table(rlogbound, paste0(outprefix, \"_rlog_normalized_Counts.gct\"), sep = \"\\t\",\n",
    "    quote = F, row.names = FALSE, col.names = FALSE, na = \"\")\n",
    "}\n",
    "\n",
    "uniqueclsclasses <- unique(coldata$condition)\n",
    "clsclasses <- coldata$condition\n",
    "clsspan <- length(rownames(coldata))\n",
    "clslen <- length(uniqueclsclasses)\n",
    "cls = data.frame(matrix(vector(), 3, clsspan, dimnames = list(c())), stringsAsFactors = F)\n",
    "cls[1, 1] <- clsspan\n",
    "cls[1, 2] <- clslen\n",
    "cls[1, 3] <- \"1\"\n",
    "cls[2, 1] <- \"# \"\n",
    "cls[2, 2:(1 + clslen)] <- uniqueclsclasses\n",
    "cls[3, ] <- clsclasses\n",
    "cat(\"Writing phenotype definitions .CLS file for GSEA.\\n\")\n",
    "write.table(cls, paste0(outprefix, \"_phenotypes.cls\"), sep = \"\\t\", quote = F, row.names = FALSE,\n",
    "  col.names = FALSE, na = \"\")\n",
    "message(\"All Done!\\n\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
